### PR TITLE
Add the QuantizedActivation linear-kernel contract

### PR DESCRIPTION
**Pull request by mgoin**
## Purpose

First part of #43224 for quant fusion support.

Introduce `QuantizedActivation`: a pre-quantized activation that fused kernel can hand directly to a linear so the linear skips its own input quant. 

The contract lives in `fusion/quant_activation.py` with two helpers:
1. `expose_input_quant_key` (the single scheme->layer bridge advertising the key a kernel can consume)
2. `as_quantized_activation` (consumer-side validate + narrow)

Linear kernels declare what they support via `input_quant_key()`; the fp8 scaled-mm and nvfp4 cutlass kernels consume the pre-quantized input, all others quantize their own.

This establishes the consumer-side contract and enforcement; the fused producers that create `QuantizedActivation`s will come separately.

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/model_executor/kernels/linear/scaled_mm/ScaledMMLinearKernel.py [hunk 1]
```diff
@@ -71,6 +75,17 @@ def __init__(self, c: _ConfigT, layer_param_names: Sequence[str]) -> None:
         self.config = c
         self.layer_param_names = layer_param_names
 
+    def input_quant_key(self) -> QuantKey | None:
```
Comment (BadrBasowid) [inline vllm/model_executor/kernels/linear/scaled_mm/ScaledMMLinearKernel.py:78]: We can instead add a static method to the interface called `apply_mm` (for lack of a better name) that bypasses the quantization operation entirely. By default, this method can raise a `NotImplementedError`, ensuring that kernels modifying the layer don't inadvertently expose it. The apply_mm method should strictly adhere to standard GEMM semantics, ensuring it behaves predictably as a pure matrix multiplication.

**Review by BadrBasowid — COMMENTED** [review]
Would like to know your opinion on the idea below (could be another clean up PR?).
Another thing worth mentioning, we probably would always want to prioritise fusion. but with this approach we'd always block fusion if the selected kernel is processing the weights, perhaps we want to skip selecting a kernel that processes the weights whenever there is an opportunity for fusion?

File: vllm/model_executor/kernels/linear/scaled_mm/ScaledMMLinearKernel.py (hunk 1 above)
Comment (BadrBasowid) [reply vllm/model_executor/kernels/linear/scaled_mm/ScaledMMLinearKernel.py:78]: this will also reduce the boilerplate if some kernels are reimplementing `apply_weights`

File: vllm/model_executor/kernels/linear/scaled_mm/ScaledMMLinearKernel.py [hunk 2]
```diff
@@ -120,30 +135,30 @@ def _get_layer_params(self, layer) -> _FP8ParamsT:
...
-        # If input not quantized
-        # TODO(luka) remove this path if not used anymore
         x_2d_q = x_2d
-        if x.dtype != fp8_dtype:
-            x_2d_q, x_s = self.quant_fp8(
-                x_2d,
-                x_s,
-                x_s_ub,
-            )
+        if x_data.dtype != fp8_dtype:
```
Comment (HDCharles) [inline vllm/model_executor/kernels/linear/scaled_mm/ScaledMMLinearKernel.py]: this is a bit hard to parse, am i understand this right, there are 3 paths for this code now?

fused-quantized (fp8), not quantized fp8, already quantized fp8? can we consolidate the branching into the qa if?

the pattern

```
qa = as_quantized_activation(x)
if qa:
   qa -> setup -> x_q
else:
   x -> quant -> x_q
# from here on out they are the same
mm(x_q,...)
```

or 

```
qa = as_quantized_activation(x)
if qa:
   qa -> setup -> x_q
elif x is not quantized:
   x -> quant -> x_q
else
   x -> x_q
# from here on out they are the same
mm(x_q,...)
```

seems clearer

File: vllm/model_executor/kernels/linear/nvfp4/flashinfer.py [hunk 3]
```diff
@@ -56,21 +68,29 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
 
-        x_fp4, x_blockscale = scaled_fp4_quant(
-            x,
-            layer.input_global_scale_inv,
-            is_sf_swizzled_layout=True,
-            backend="flashinfer-cutlass",
-            padded_n=x.shape[-1] + weights_padding_bytes * 2,
-        )
+        qa = as_quantized_activation(x, self.input_quant_key())
+        if qa is not None:
```
Comment (HDCharles) [inline vllm/model_executor/kernels/linear/nvfp4/flashinfer.py:79]: sorry if this is a dumb question, but why this as_quantized_activation api instead of something simpler:

```
if is_quantized_activation(x, self.input_quant_key()):
    x -> setup -> x_q
else:
    x -> quant -> x_q
```

is it the variable name change so one branch uses qa and another uses x?

File: vllm/model_executor/layers/fusion/quant_activation.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+import torch
+
+from vllm.model_executor.layers.quantization.utils.quant_utils import QuantKey
+
+
+@dataclass
+class QuantizedActivation:
+    data: torch.Tensor
+    scale: torch.Tensor
```
Comment (HDCharles) [inline vllm/model_executor/layers/fusion/quant_activation.py:32]: i assume this means the fusion is impossible if there's a zero point?

File: vllm/model_executor/kernels/linear/nvfp4/flashinfer.py [hunk 5]
```diff
@@ -56,21 +68,29 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
+            x_fp4, x_blockscale = qa.data, qa.scale
+            x_fp4 = pad_nvfp4_activation_for_cutlass(x_fp4, weights_padding_bytes)
```
Comment (zyongye) [inline vllm/model_executor/kernels/linear/nvfp4/flashinfer.py:81]: should we add padding/alignment/layout requirement into quantkey as well? So that we can save extra kernel launch for this

File: vllm/model_executor/layers/fusion/quant_activation.py (hunk 4 above)
Comment (mgoin) [reply vllm/model_executor/layers/fusion/quant_activation.py:32]: We don't have any fused kernels that use zero points (basically fp8, nvfp4 only), so no need to represent that yet

File: vllm/model_executor/kernels/linear/nvfp4/flashinfer.py (hunk 5 above)
Comment (mgoin) [reply vllm/model_executor/kernels/linear/nvfp4/flashinfer.py:81]: This is a later concern and I'm open to it. Ultimately we aren't regressing anything here since we don't have fused kernels available for the unique trtllm swizzled layout or padded activation. 

It also brings up a hard tradeoff to make: if you have an all-reduce+rms+quant kernel that doesn't support padding versus an rms+quant that supports padding, which one do you choose in the best case?

File: vllm/model_executor/kernels/linear/scaled_mm/ScaledMMLinearKernel.py (hunk 1 above)
Comment (mgoin) [reply vllm/model_executor/kernels/linear/scaled_mm/ScaledMMLinearKernel.py:78]: The branch must live somewhere, and only the kernel knows its own consumption prep (nvfp4 re-pad, shape/dtype recovery); so "strict GEMM semantics" would push kernel-specific logic to a caller that can't know it, IIUC.

File: vllm/model_executor/kernels/linear/nvfp4/flashinfer.py (hunk 3 above)
Comment (mgoin) [reply vllm/model_executor/kernels/linear/nvfp4/flashinfer.py:79]: Yeah ultimately it is just a style choice and wanting the hide the repeated boilerplate a bit. It is trivial to change in the future. I don't feel strongly either way
